File: Month00_CSharp/Workers/Incoming/Work.cs
using System;
using System.Collections.Generic;

namespace Workers.Incoming;

internal readonly record struct Work
{
    private static readonly ExecutionModel[] ExecutionModels = Enum.GetValues<ExecutionModel>();
    private static readonly Behaviour[]      Behaviours      = Enum.GetValues<Behaviour>();

    public int Id { get; private init; }

    public ExecutionModel ExecutionModel { get; private init; }

    public Behaviour Behaviour { get; private init; }

    public static Work Create(int id) =>
        new()
        {
            Id             = id,
            ExecutionModel = PickRandomExecutionModel(),
            Behaviour      = PickRandomBehaviour()
        };

    private static ExecutionModel PickRandomExecutionModel() =>
        ExecutionModels[Random.Shared.Next(ExecutionModels.Length)];

    private static Behaviour PickRandomBehaviour()
    {
        ulong mask    = 0;
        var   singles = new List<byte>(Behaviours.Length);

        foreach (var value in Behaviours)
        {
            var bits = (byte)value;

            if (bits == 0)
            {
                continue;
            }

            // Only treat power-of-two values as "atomic" flags.
            if ((bits & (bits - 1)) != 0)
            {
                continue;
            }

            singles.Add(bits);

            if (Random.Shared.Next(2) == 0)
            {
                mask |= bits;
            }
        }

        // Ensure we don't end up with "None" (0) too often.
        if (mask == 0 && singles.Count > 0)
        {
            mask = singles[Random.Shared.Next(singles.Count)];
        }

        return (Behaviour)mask;
    }
}

File: 03_FSharp/Workers/Incoming/Work.fs
namespace Workers.Incoming

open System
open Workers.Incoming

[<Struct>]
type Work =
    { Id: int
      ExecutionModel: ExecutionModel
      Behaviour: Behaviour }

    static let executionModels: ExecutionModel[] = Enum.GetValues<ExecutionModel>()

    static let behaviours: Behaviour[] = Enum.GetValues<Behaviour>()

    static member public Create(id: int) : Work =
        { Id = id
          ExecutionModel = Work.PickRandomExecutionModel()
          Behaviour = Work.PickRandomBehaviour() }

    static member private PickRandomExecutionModel() : ExecutionModel =
        executionModels[Random.Shared.Next(executionModels.Length)]

    static member private PickRandomBehaviour() : Behaviour =
        let singles: byte[] =
            behaviours
            |> Array.choose (fun value ->
                let b = byte value

                if b = 0uy then
                    None
                else
                    let x = int b
                    if (x &&& (x - 1)) = 0 then Some b else None)

        let mask: byte =
            singles
            |> Array.fold (fun acc b -> if Random.Shared.Next(2) = 0 then acc ||| b else acc) 0uy

        // Ensure we don't end up with "None" (0) too often.
        let finalMask =
            if mask = 0uy && singles.Length > 0 then
                singles[Random.Shared.Next(singles.Length)]
            else
                mask

        LanguagePrimitives.EnumOfValue finalMask
